### PR TITLE
Fixes #10466 - nice error on proxy registration error

### PR DESCRIPTION
**Pull request by lzap**
Opinions on warnings instead of errors? Since registration step is not fatal
and it often confuses users I'd vote for this. Otherwise I can just raise an
error with `raise "bla bla"`

### CODE REVIEW

**Comment by lzap** [comment]
To reproduce just send invalid name/url or simply rename existing proxy name to something different and re-run with the same URL.

**Comment by domcleal** [comment]
This should still raise an error, as a resource that doesn't actually fail when it meets an error is really quite useless...

File: lib/puppet/provider/foreman_smartproxy/rest_v2.rb [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
   confine :feature => :apipie_bindings
 
+  def print_warning(e)
+    body = JSON.parse(e.response)["error"]["full_messages"].join(" ") rescue 'N/A'
+    warning("Error while registering proxy (#{e.message}): #{body}, " +
+      "registration can be skipped by unsetting --register-in-foreman option.")
```
Comment (domcleal) [inline lib/puppet/provider/foreman_smartproxy/rest_v2.rb]: This message is rather tied to the module.. where the parameter is `register_in_foreman`, or if you meant the foreman-installer option, then it's `--foreman-register-in-foreman`.

**Comment by lzap** [comment]
Yeah sure. Fixed.

File: lib/puppet/provider/foreman_smartproxy/rest_v2.rb [hunk 2]
```diff
@@ -2,6 +2,12 @@
 
   confine :feature => :apipie_bindings
 
+  def raise_error(e)
+    body = JSON.parse(e.response)["error"]["full_messages"].join(" ") rescue 'N/A'
+    raise("Error while registering proxy (#{e.message}): #{body}, " +
```
Comment (domcleal) [inline lib/puppet/provider/foreman_smartproxy/rest_v2.rb]: Better to use the builtin fail() in a provider

File: lib/puppet/provider/foreman_smartproxy/rest_v2.rb [hunk 3]
```diff
@@ -2,6 +2,12 @@
...
+      "registration can be skipped by unsetting --register-in-foreman option.")
```
Comment (domcleal) [inline lib/puppet/provider/foreman_smartproxy/rest_v2.rb]: see my last comment about the name of the parameter

Comment (domcleal) [inline lib/puppet/provider/foreman_smartproxy/rest_v2.rb]: Could we emphasise finding the reason before disabling it too, so people don't follow this advice and get a partially functioning installation?  IIRC we have debug output from apipie-bindings nowadays.

File: lib/puppet/provider/foreman_smartproxy/rest_v2.rb (hunk 2 above)
Comment (lzap) [inline lib/puppet/provider/foreman_smartproxy/rest_v2.rb]: Fixed.

File: lib/puppet/provider/foreman_smartproxy/rest_v2.rb (hunk 3 above)
Comment (lzap) [inline lib/puppet/provider/foreman_smartproxy/rest_v2.rb]: Re: name of parameter - I don't see any related comment on that.

Can you suggest wording you expect? I am really bad in this... I don't follow what debug output from bindings you mean.

File: lib/puppet/provider/foreman_smartproxy/rest_v2.rb [hunk 4]
```diff
@@ -2,6 +2,12 @@
 
   confine :feature => :apipie_bindings
 
+  def raise_error(e)
+    body = JSON.parse(e.response)["error"]["full_messages"].join(" ") rescue 'N/A'
+    fail("Error while registering proxy (#{e.message}): #{body}, " +
+      "registration can be skipped by unsetting --register-in-foreman option.")
```
Comment (ekohl) [inline lib/puppet/provider/foreman_smartproxy/rest_v2.rb]: This is only relevant for the installer. Can we come up with something that also works in all cases? I'd prefer if all users registered the proxy so I wonder if we should even encourage users to disable registration.

**Comment by ekohl** [comment]
I think the more detailed error can be very helpful so yay for the idea.

**Comment by lzap** [comment]
Ok I've simplified the error: `fail "Proxy #{resource[:name]} cannot be registered (#{e.message}): #{body}"`

**Comment by ekohl** [comment]
:+1: and I think the test is unrelated (and was already broken before).

**Comment by domcleal** [comment]
Looks better, thanks.  There are a couple of other methods at the bottom which should perhaps be guarded too - url= and refresh_features.

**Comment by lzap** [comment]
Oh thanks for help. We should be all set.

**Comment by mmoll** [comment]
very nice, thanks @lzap!

**Pull request merged**